contracts/multisig-prover/src/contract.rs

```diff
@@ -1,44 +1,479 @@
+    }
+
+    pub fn rotate_snapshot(
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        pub_keys: HashMap<String, HexBinary>,
+        key_id: String,
+    ) -> Result<Response, ContractError> {
+        KEY_ID.save(deps.storage, &key_id)?;
+
+        let config = CONFIG.load(deps.storage)?;
+        if config.admin != info.sender {
+            return Err(ContractError::Unauthorized {});
+        }
+
+        let snapshot = snapshot(deps.querier, env.block, &config)?;
+
+        let keygen_msg = WasmMsg::Execute {
+            contract_addr: config.multisig.into(),
+            msg: to_binary(&multisig::msg::ExecuteMsg::KeyGen {
+                key_id: key_id.clone(),
+                snapshot: snapshot.clone(),
+                pub_keys: pub_keys.clone(),
+            })?,
+            funds: vec![],
+        };
+
+        let event = Event::SnapshotRotated {
+            key_id,
+            snapshot,
+            pub_keys,
+        };
+
+        Ok(Response::new()
+            .add_message(keygen_msg)
+            .add_event(event.into()))
+    }
+
+    fn snapshot(
+        querier: QuerierWrapper,
+        block: BlockInfo,
+        config: &Config,
+    ) -> Result<Snapshot, ContractError> {
+        let query_msg = service_registry::msg::QueryMsg::GetActiveWorkers {
+            service_name: config.service_name.clone(),
+            chain_name: config.chain_name.to_string(),
+        };
+
+        let active_workers: Vec<Worker> = querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
+            contract_addr: config.service_registry.to_string(),
+            msg: to_binary(&query_msg)?,
+        }))?;
+
+        let participants = active_workers
+            .into_iter()
+            .map(Worker::try_into)
+            .collect::<Result<Vec<Participant>, service_registry::ContractError>>()
+            .map_err(
+                |err: service_registry::ContractError| ContractError::InvalidParticipants {
+                    reason: err.to_string(),
+                },
+            )?
+            .try_into()
+            .map_err(|err: axelar_wasm_std::nonempty::Error| {
+                ContractError::InvalidParticipants {
+                    reason: err.to_string(),
+                }
+            })?;
+
+        Ok(Snapshot::new(
+            block
+                .time
+                .try_into()
+                .expect("violated invariant: block time cannot be invalid"),
+            block
+                .height
+                .try_into()
+                .expect("violated invariant: block height cannot be invalid"),
+            config.signing_threshold,
+            participants,
+        ))
     }
 }
 
 #[cfg_attr(not(feature = "library"), entry_point)]
-pub fn query(_deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+pub fn reply(deps: DepsMut, _env: Env, reply: Reply) -> Result<Response, ContractError> {
+    match parse_reply_execute_data(reply) {
```

Comment:
Should match the `reply.id` instead to see it it equals `START_MULTISIG_REPLY_ID`